### PR TITLE
Reading characteristics for id, service, device

### LINKED ISSUE #19: Reading characteristic from peripheral/service/characteristic

### PR DESCRIPTION
**Pull request by mikolak**
Resolve #19

### CODE REVIEW

File: src/internal/delegates/characteristics-delegate.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        characteristicIdentifier: number): Promise<TransferCharacteristic | SimulatedBleError> {
+        try {
+            let matchedPeripheral: SimulatedPeripheral | null = null
+            for (let i = 0; i < peripherals.length; i++) {
+                if (peripherals[i].getCharacteristic(characteristicIdentifier) != null) {
+                    matchedPeripheral = peripherals[i];
+                    break
+                }
+            }
+
```
Comment (mikolak) [inline src/internal/delegates/characteristics-delegate.ts:20]: Add checks on adapter state

File: src/internal/delegates/characteristics-delegate.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import { SimulatedCharacteristic } from "../../simulated-characteristic";
+import { TransferCharacteristic, mapToTransferCharacteristic } from "../internal-types";
+
+export class CharacteristicsDelegate {
+    async readCharacteristic(adapterState: AdapterState,
+        peripherals: Array<SimulatedPeripheral>,
+        characteristicIdentifier: number): Promise<TransferCharacteristic | SimulatedBleError> {
+        try {
+            let matchedPeripheral: SimulatedPeripheral | null = null
+            for (let i = 0; i < peripherals.length; i++) {
```
Comment (pawelByszewski) [inline src/internal/delegates/characteristics-delegate.ts:14]: IMHO extracting that to separate method would increase readability.

File: src/internal/delegates/characteristics-delegate.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        }
+    }
+
+    async readCharacteristicForService(adapterState: AdapterState,
+        peripherals: Array<SimulatedPeripheral>,
+        serviceIdentifier: number,
+        characteristicUuid: UUID): Promise<TransferCharacteristic | SimulatedBleError> {
+        try {
+            let matchedPeripheral: SimulatedPeripheral | null = null
+            for (let i = 0; i < peripherals.length; i++) {
```
Comment (pawelByszewski) [inline src/internal/delegates/characteristics-delegate.ts:49]: IMHO extracting that to separate method would increase readability.

File: src/internal/delegates/characteristics-delegate.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        try {
+            let matchedPeripheral: SimulatedPeripheral | undefined = peripherals.get(peripheralIdentifier)
+
+            errorIfPeripheralNotFound(matchedPeripheral)
+            errorIfPeripheralNotConnected(matchedPeripheral!)
+            errorIfDiscoveryNotDone(matchedPeripheral!)
+
+            let characteristic: SimulatedCharacteristic | undefined = matchedPeripheral!.getCharacteristicForService(serviceUuid, characteristicUuid)
+            errorIfCharacteristicNotFound(characteristic)
+            errorIfNotReadable(characteristic!)
```
Comment (pawelByszewski) [inline src/internal/delegates/characteristics-delegate.ts:97]: it seems next four lines are copied the third time.

**Pull request merged**